chromium_src/chrome/browser/printing/print_preview_message_handler.cc

```diff
@@ -0,0 +1,235 @@
+  scoped_refptr<printing::PrinterQuery> printer_query =
+      queue->PopPrinterQuery(document_cookie);
+  if (printer_query.get()) {
+    BrowserThread::PostTask(BrowserThread::IO, FROM_HERE,
+                            base::Bind(&printing::PrinterQuery::StopWorker,
+                                       printer_query));
+  }
+}
+
+base::RefCountedBytes* GetDataFromHandle(base::SharedMemoryHandle handle,
+                                         uint32 data_size) {
+  scoped_ptr<base::SharedMemory> shared_buf(
+      new base::SharedMemory(handle, true));
+  if (!shared_buf->Map(data_size)) {
+    NOTREACHED();
+    return NULL;
+  }
+
+  unsigned char* data_begin = static_cast<unsigned char*>(shared_buf->memory());
+  std::vector<unsigned char> data(data_begin, data_begin + data_size);
+  return base::RefCountedBytes::TakeVector(&data);
+}
+
+printing::PrintPreviewMessageHandler::PrintPDFResult
+SavePDF(const scoped_refptr<base::RefCountedBytes>& data,
+        const base::FilePath& save_path) {
+  printing::PdfMetafileSkia metafile;
+  metafile.InitFromData(static_cast<const void*>(data->front()), data->size());
+  base::File file(save_path,
+                  base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE);
+  return metafile.SaveTo(&file)?
+    printing::PrintPreviewMessageHandler::SUCCESS :
+    printing::PrintPreviewMessageHandler::FAIL_SAVEFILE;
+}
+
+}  // namespace
+
+namespace printing {
+
+
+PrintPreviewMessageHandler::PrintPreviewMessageHandler(
+    WebContents* web_contents)
+    : content::WebContentsObserver(web_contents),
+      request_id_(0) {
+  DCHECK(web_contents);
+}
+
+PrintPreviewMessageHandler::~PrintPreviewMessageHandler() {
+}
+
+void PrintPreviewMessageHandler::OnDidGetPreviewPageCount(
+    const PrintHostMsg_DidGetPreviewPageCount_Params& params) {
+  if (params.page_count <= 0) {
+    NOTREACHED();
+    return;
+  }
+}
+
+void PrintPreviewMessageHandler::OnDidPreviewPage(
+    const PrintHostMsg_DidPreviewPage_Params& params) {
+  int page_number = params.page_number;
+  if (page_number < FIRST_PAGE_INDEX || !params.data_size)
+    return;
+}
+
+void PrintPreviewMessageHandler::OnMetafileReadyForPrinting(
+    const PrintHostMsg_DidPreviewDocument_Params& params) {
+  // Always try to stop the worker.
+  StopWorker(params.document_cookie);
+
+  if (params.expected_pages_count <= 0) {
+    NOTREACHED();
+    return;
+  }
+
+  // TODO(joth): This seems like a good match for using RefCountedStaticMemory
+  // to avoid the memory copy, but the SetPrintPreviewData call chain below
+  // needs updating to accept the RefCountedMemory* base class.
+  scoped_refptr<base::RefCountedBytes> data(
+      GetDataFromHandle(params.metafile_data_handle, params.data_size));
+  if (!data || !data->size())
+    return;
+
+  atom::NativeWindow* window = atom::NativeWindow::FromWebContents(
+      web_contents());
+  base::FilePath save_path;
+  if (!file_dialog::ShowSaveDialog(window, "Save As",
```

Comment:
The filename should be passed as a parameter of `printToPdf`, users might want to do the printing silently.